flang/docs/ParallelFortranRuntime.md

```diff
@@ -0,0 +1,2559 @@
+  team. If `team` is not present, the current team is used.
+    
+```
+subroutine prif_team_number(...)
+  type(prif_team_type), intent(in), optional :: team
+  integer(c_intmax_t), intent(out) :: team_number
+end subroutine
+```
+
+### `prif_change_team`
+
+**Description**: changes the current team to the specified team. For any
+  associate names specified in the `CHANGE TEAM` statement the compiler should
+  follow a call to this procedure with calls to `prif_alias_create` to create
+  an alias coarray descriptor, and associate any non-coindexed references to the
+  associate name within the `CHANGE TEAM` construct with the selector.
+    
+```
+subroutine prif_change_team(...)
+  type(prif_team_type), intent(in) :: team
+  integer(c_int), intent(out), optional :: stat
+  character(len=*), intent(inout), optional :: errmsg
+  character(len=:), intent(inout), allocatable, optional :: errmsg_alloc
+end subroutine
+```
+
+### `prif_end_team`
+
+**Description**: Changes the current team to the parent team. During the
+  execution of `prif_end_team`, the PRIF implementation will deallocate any coarrays that became allocated during the
+  change team construct. Prior to invoking `prif_end_team`, the compiler is
+  responsible for invoking `prif_alias_destroy` to delete any coarray alias descriptors
+  created as part of the `CHANGE TEAM` construct.
+    
+```
+subroutine prif_end_team(...)
+  integer(c_int), intent(out), optional :: stat
+  character(len=*), intent(inout), optional :: errmsg
+  character(len=:), intent(inout), allocatable, optional :: errmsg_alloc
+end subroutine
+```
+
+## Collectives
+
+### Common Arguments in Collectives
+
+* **`a`**
+  * Argument for all the collective subroutines: [`prif_co_broadcast`](#prif_co_broadcast),
+    [`prif_co_max`](#prif_co_max), [`prif_co_min`](#prif_co_min),
+    [`prif_co_sum`](#prif_co_sum), [`prif_co_reduce`](#prif_co_reduce).
+  * may be any type for `prif_co_broadcast` or `prif_co_reduce`, any numeric for `prif_co_sum`,
+    and integer, real, or character for `prif_co_min` or `prif_co_max`
+  * is always `intent(inout)`
+  * for `prif_co_max`, `prif_co_min`, `prif_co_sum`, `prif_co_reduce` it is assigned the value
+    computed by the collective operation, if no error conditions occurs and if
+    `result_image` is absent, or the executing image is the one identified by
+    `result_image`, otherwise `a` becomes undefined
+  * for `prif_co_broadcast`, the value of the argument on the `source_image` is
+    assigned to the `a` argument on all other images
+
+* **`source_image` or `result_image`**
+  * Identifies the image in the current team that is the root of the collective operation.
+  * If `result_image` is omitted, then all participating images receive the resulting value.
+
+### `prif_co_broadcast`
+
+**Description**: Broadcast value to images
+    
+```
+subroutine prif_co_broadcast(...)
+  type(*), intent(inout), contiguous, target :: a(..)
```

Comment:
@jeffhammond's comment:
> Can you remind me why collective buffer arguments are contiguous, target? Is there any assumption that the compiler copies non-contiguous subarrays here? What is the target attribute doing? Thanks.

**LBL response:** @jeffhammond The rationale for the `contiguous` and `target` attributes on the argument `a` is so that a PRIF implementation can assume contiguous and point at the buffer.

The `contiguous` attribute should allow the compiler to elide copies where possible but it will need to make copies in the case that they are non-contiguous subarrays. 

The `target` attribute puts no requirements on the actual argument but it allows a PRIF implementation to have a pointer to it.

Reply:
Why not let the compiler test for contiguousness and not force copies for noncontiguous cases?  The branch is trivial.  The copy is not.

What is the purpose of taking a pointer to the buffer?  When is the implementation going to create a pointer to this buffer, versus use its address directly?

Reply:
This is resolved in PRIF 0.5, which removes the `contiguous` attribute from the `a` argument to collective subroutines.